src/java.naming/share/classes/com/sun/jndi/ldap/LdapRequest.java

```diff
@@ -61,87 +60,94 @@ final class LdapRequest {
+        try {
+            replies.put(ber);
+        } catch (InterruptedException e) {
+            // ignore
         }
+        return pauseAfterReceipt;
     }
 
     /**
      * Read reply BER
      * @param millis timeout, infinite if the value is negative
      * @return BerDecoder if reply was read successfully
-     * @throws CommunicationException request has been canceled and request does not need to be abandoned
-     * @throws IOException            request has been closed or timed out. Request does need to be abandoned
-     * @throws InterruptedException   LDAP operation has been interrupted
+     * @throws CommunicationException request has been canceled and request
+     *                                does not need to be abandoned (i.e. a LDAP_REQ_ABANDON
+     *                                message need not be sent across)
+     * @throws IOException            request has been closed or timed out.
+     *                                Request needs to be abandoned (i.e. a LDAP_REQ_ABANDON
+     *                                message needs to be sent across)
+     * @throws InterruptedException   the wait to read a reply has been interrupted
      */
+    // more than one thread invoking this method concurrently isn't expected
     BerDecoder getReplyBer(long millis) throws IOException, CommunicationException,
                                                InterruptedException {
-        if (cancelled) {
-            throw new CommunicationException("Request: " + msgId +
-                " cancelled");
-        }
-        if (isClosed()) {
-            throw new IOException(CLOSE_MSG);
-        }
-
-        BerDecoder result = millis > 0 ?
-                replies.poll(millis, TimeUnit.MILLISECONDS) : replies.take();
 
-        if (cancelled) {
-            throw new CommunicationException("Request: " + msgId +
-                " cancelled");
+        final boolean hasReplies = replies.peek() != null;
+        if (!hasReplies) {
+            // no replies have been queued, so if the request has
+            // been cancelled or closed, then raise an exception
+            if (cancelled) {
+                throw new CommunicationException("Request: " + msgId +
+                        " cancelled");
+            }
+            if (closed) {
+                throw new IOException(CLOSE_MSG);
+            }
         }
-
+        // either there already are queued replies or the request is still
+        // alive (i.e. not cancelled or closed). we wait for a reply to arrive
+        // or the request to be cancelled/closed, in which case the replies
+        // queue will contain the relevant marker.
+        final BerDecoder result = millis > 0
+                ? replies.poll(millis, TimeUnit.MILLISECONDS)
+                : replies.take();
         // poll from 'replies' blocking queue ended-up with timeout
         if (result == null) {
             throw new IOException(String.format(TIMEOUT_MSG_FMT, millis));
```

Comment:
```suggestion
            throw new IOException("LDAP response read timed out, timeout used: %d ms.".formatted(millis));
```
TIMEOUT_MSG_FMT is only used once here, and defining a constant would make the code less readable.